core/src/kmx/kmx_plus.cpp

```diff
@@ -277,34 +281,109 @@ COMP_KMXPLUS_SECT::valid(KMX_DWORD length) const {
     KMX_DWORD entrylength = length - entry.offset;
     // just validate header
     if(header_from_bytes(entrydata, entrylength, entry.sect) == nullptr) {
-      KMXPLUS_PRINTF(("Invalid header %X", entry.sect));
+      DebugLog("Invalid header %X", entry.sect);
       overall_valid = false;
       continue;
     }
   }
   return overall_valid;
 }
 
+// ---- transform related fcns
+bool
+COMP_KMXPLUS_ELEM::valid(KMX_DWORD _kmn_unused(length)) const {
+  if (header.size < sizeof(*this)+(sizeof(entries[0])*count)) {
+    DebugLog("header.size 0x%X < expected size");
+    return false;
+  }
+  const COMP_KMXPLUS_ELEM_ENTRY &firstEntry = entries[0];
+  if (firstEntry.length != 0 ) {
+    // per spec, the first elem should have zero length and zero offset
+    DebugLog("ERROR: elem[0].length 0x%x but should be zero", firstEntry.length);
+    return false;
+  }
+  if (firstEntry.offset + sizeof(COMP_KMXPLUS_ELEM_ELEMENT) > header.size) {
+    // TODO-LDML: change to  (firstEntry.offset != 0 )
+    // Blocked by https://github.com/keymanapp/keyman/issues/7404
+    DebugLog("WARNING: elem[0].offset 0x%x would put element outside of data region..."
+           "Please fix https://github.com/keymanapp/keyman/issues/7404");
+    // TODO-LDML: should exit here
+    // return false;
+  }
+  for (KMX_DWORD e = 1; e < count; e++) {
+    // Don't need to recheck the first entry hbere.
+    KMX_DWORD listLength;
+    if (getElementList(e, listLength) == nullptr) {
+      return false;
+    }
+    if (listLength == 0) { // only the first element should have length zero
+      DebugLog("ERROR: elem[0x%x].length == 0", e);
+      return false;
+    }
+  }
+  return true;
+}
+
+const COMP_KMXPLUS_ELEM_ELEMENT *
+COMP_KMXPLUS_ELEM::getElementList(KMX_DWORD elementNumber, KMX_DWORD &length) const {
+  if (elementNumber >= count) {
+    return nullptr;
```

Comment:
This should have a DebugLog statement and/or debug-build assertion.

Reply:
done